flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java

```diff
@@ -104,11 +116,216 @@ public void testHandleMixedLargeRecords() throws Exception {
 		testSerializationRoundTrip(originalRecords, segmentSize);
 	}
 
+	/**
+	 * Non-spanning, deserialization reads one byte too many and succeeds.
+	 */
+	@Test
+	public void testHandleDeserializingTooMuchNonSpanning1() throws Exception {
+		expectedException.expect(IOException.class);
+		expectedException.expectMessage(" -1 remaining unread byte");
+		testHandleWrongDeserialization(new StringValueDeserializingTooMuch("Test string"), 32 * 1024);
+	}
+
+	/**
+	 * Non-spanning, serialization length is 16 (including headers), deserialization reads one byte
+	 * too many and succeeds.
+	 */
+	@Test
+	public void testHandleDeserializingTooMuchNonSpanning2() throws Exception {
+		expectedException.expect(IOException.class);
+		expectedException.expectMessage(" -1 remaining unread byte");
+		testHandleWrongDeserialization(new StringValueDeserializingTooMuch("Test string"), 17);
+	}
+
+	/**
+	 * Non-spanning, serialization length is 16 (including headers), deserialization reads one byte
+	 * too many and fails.
+	 */
+	@Test
+	public void testHandleDeserializingTooMuchNonSpanning3() throws Exception {
+		expectedException.expect(IOException.class);
+		expectedException.expectMessage(" -1 remaining unread byte");
+		expectedException.expectCause(isA(IndexOutOfBoundsException.class));
+		testHandleWrongDeserialization(new StringValueDeserializingTooMuch("Test string"), 16);
+	}
+
+	/**
+	 * Spanning, serialization length is 16 (including headers), deserialization reads one byte
+	 * too many and fails.
+	 */
+	@Test
+	public void testHandleDeserializingTooMuchSpanning1() throws Exception {
+		expectedException.expect(IOException.class);
+		expectedException.expectMessage(" -1 remaining unread byte");
+		expectedException.expectCause(isA(EOFException.class));
+		testHandleWrongDeserialization(new StringValueDeserializingTooMuch("Test string"), 15);
+	}
+
+	/**
+	 * Spanning, serialization length is 16 (including headers), deserialization reads one byte
+	 * too many and fails.
+	 */
+	@Test
+	public void testHandleDeserializingTooMuchSpanning2() throws Exception {
```

Comment:
Also this test can be merged with above `testHandleDeserializingTooMuchSpanning1` because of same `expectedException`.